fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/ContainmentIndexImpl.java

```diff
@@ -278,24 +313,55 @@ public void removeContainedBy(final Transaction tx, final FedoraResource parent,
         }
     }
 
+    @Override
+    public void removeResource(final String txID, final String resourceID) {
+        final MapSqlParameterSource parameterSource = new MapSqlParameterSource();
+        parameterSource.addValue("child", resourceID);
+        if (txID != null) {
+            parameterSource.addValue("transactionId", txID);
+            final boolean addedInTxn = !jdbcTemplate.queryForList(IS_CHILD_ADDED_IN_TRANSACTION_NO_PARENT,
+                    parameterSource).isEmpty();
+            if (addedInTxn) {
+                jdbcTemplate.update(UNDO_INSERT_CHILD_IN_TRANSACTION_NO_PARENT, parameterSource);
+            } else {
+                final String parent = getContainedBy(txID, resourceID);
+                if (parent != null) {
+                    LOGGER.debug("Removing containment relationship between parent ({}) and child ({})", parent,
+                            resourceID);
+                    parameterSource.addValue("parent", parent);
+                    jdbcTemplate.update(DELETE_CHILD_IN_TRANSACTION, parameterSource);
+                }
+            }
+        } else {
+            jdbcTemplate.update(DELETE_ALL_RESOURCE, parameterSource);
```

Comment:
Isn't this removing the resource from the actual containment index, rather than staging it to be removed with the transaction is committed? Is this what we want?

Reply:
If there is no transaction then this is the path. We could also not allow this to be used without a transaction, but I was following the existing semantics. In my mind, this probably would never happen, but I don't know.

Reply:
Ah, fair enough.